### DISCUSSION TITLE
Prisma on node.js cpanel shared hosting

### DISCUSSION THREAD

**Discussion by pablopsp** [prisma/prisma#6647] [Q&A] [answered]
I'm trying to deploy my express + prisma backend in a Cpanel located on a shared hosting and i'm facing this error.

```
(node:1083) UnhandledPromiseRejectionWarning: Error: Query engine exited with code 127

/home/cmp/nodevenv/api/12/lib/node_modules/.prisma/client/query-engine-debian-openssl-1.1.x: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
    at cb (/home/cmp/nodevenv/api/12/lib/node_modules/@prisma/client/runtime/index.js:34786:17)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
(node:1083) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:1083) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

I'm generating the prisma client with `npx prisma generate` from my local env then uploading the `node_modules` to the server, because when I use the script from the server I got this error:

```console
Prisma schema loaded from prisma/schema.prisma
stderr:
Error: Get config /home/cmp/.npm/_npx/62793/lib/node_modules/prisma/query-engine-debian-openssl-1.1.x: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
```  

My `schema.prisma` file:
```prisma
generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "debian-openssl-1.1.x"]
}

datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
}

//Models..
```

What can I do to make it works as it should?

> Versions

`prisma: 2.21.2`
`@prisma/client: 2.21.2`
`node -v: v12.21.0`

**Comment by 2color** [comment, marked as answer]
Hey @pablopsp,
Do you know which operating system you're running in? 

The error you're experiencing indicates that you don't have openssl 1.1 on your machine which is a requirement for the Prisma engine.

I would double-check which OS version you are running and update the `binaryTargets` in your Prisma schema based on the [following docs](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference/#binarytargets-options)

**Reply by pablopsp** [reply]
Hey, thanks for your fast reply.
I think that is  CloudLinux, `uname -a` gives me :
`Linux 3.10.0-962.3.2.lve1.5.42.el7.x86_64 #1 SMP Mon Nov 9 08:11:18 EST 2020 x86_64 x86_64 x86_64 GNU/Linux`

I saw that my server has `OpenSSL 1.0.2k-fips`, it's possible to work with that version?
I tried with one of the `1.0.x` engine listed on the docs for the`binaryTargets`  and I got an error that says:
`Error: Query engine binary for current platform "debian-openssl-1.1.x" could not be found.`

**Comment by kenzeasy** [comment]
Hey @pablopsp, sadly not an answer to your issue but I got something quite similar to yours. I was about to open a discussion here when I saw yours. 

I'm currently trying to make Prisma works on a shared hosting through a node app and Cpanel. 
As you when I run the `npx prisma migrate dev --name init` or `npx prisma generate` or basically any other npx prisma commands (except npx prisma init) I get the same error as you : "error while loading shared lib" I've tried a bunch of stuff to make this work and so far none are working. 

I've seen this "libssl.so.1.1" sitting somewhere on the shared hosting but the node app seems to not have access to it 

By contacting the support of my shared hosting I know that they are currently running Cent OS 7 and I've tried to add this specific binary target on my prisma.schema file and it's not changing anything.

As @2color mentioned it, I've checked the version of the openssl installed on the shared hosting and as you it is the 1.0.2k-fips so it might be the key to our trouble here ... I don't know if there is any workaround or I need to deploy a prisma app and a DB external to this specific shared hosting because openssl is not up to date ...

**Reply by janpio** [reply]
Can you share the schema you tried to configure for that environment @kenzeasy?
Can you also share the shared web host? Maybe someone wants to reproduce your experience to help us figure out what could be wrong.

**Reply by kenzeasy** [reply]
Sure the shared hosting is a french one I've tried to find their "international" page but I could not : https://www.o2switch.fr/

I'm following the [start from scratch](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch-node-postgres) my prisma.schema is a copy pasta from it I've only changed the generator client part to match the Cent OS 7: 
```
// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
  binaryTargets = ["rhel-openssl-1.0.x"]
}

//Models...
```
It rise some questions tho since I know that the missing lib "libssl.so.1.1" is nested somewhere into the root folder of the shared hosting (I've tried to copy it in different places to see if finally the app gain access to it but without success...) but the virtual environment setup seems to not have access to it even if openssl is not 1.1 version...

**Comment by kenzeasy** [comment]
Some updates even if I know that it might be complicated to replicate (or maybe not idk...) but I do want to make Prisma works and since there is a User table in my postgres DB on my shared hosting I guess I've done it ... I did not tried anything else yet.

**Please let me know if there is something that I've done that is not recommanded, beyond my usual ape state, or out of this mortal world.** 

Here's my path : 

I've started again and again the [start from scratch](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch-node-postgres) tuto everytime I I've found some new infos and yesterday I saw this discussion post [migration engine download in build pipeline](https://github.com/prisma/prisma/discussions/5965) which provide a url to download the binaries that might be needed by the shared hosting OS (In my case Cent OS 7 if I can trust the support that I've contacted).

 So ... I've tried ! 

I've changed the "commit hash string" (I don't know how this string is called) to match the one that I've found inside package.json from a local fresh install of Prisma and the last part of the url to match the different binaries and downloaded all the files. You can see the urls used in the links below.

[query engine](https://binaries.prisma.sh/all_commits/e421996c87d5f3c8f7eeadd502d4ad402c89464d/rhel-openssl-1.0.x/query-engine.gz)
[introspection engine](https://binaries.prisma.sh/all_commits/e421996c87d5f3c8f7eeadd502d4ad402c89464d/rhel-openssl-1.0.x/introspection-engine.gz)
[migration engine](https://binaries.prisma.sh/all_commits/e421996c87d5f3c8f7eeadd502d4ad402c89464d/rhel-openssl-1.0.x/migration-engine.gz)
[prisma-fmt](https://binaries.prisma.sh/all_commits/e421996c87d5f3c8f7eeadd502d4ad402c89464d/rhel-openssl-1.0.x/prisma-fmt.gz)

I've created a new nodejs app on the cpanel shared hosting and followed again the tuto  
```
// this comes from the shared hosting to get access to node, npm, npx ... 
source ~/nodevenv/myapp/14/bin/activate && cd ~/myapp

// this comes from the prisma tuto
npm init -y

npm install prisma --save-dev

npx prisma init
```
From there as usual Prisma is downloading all the `debian-openssl-1.1.x` query, migration, introspection and prisma-fmt binaries since it assume that the OS is debian based. But I knew that will not work and I'll face again a "get config error : openssl 1.1 is missing".

So I've extracted from the archives the "new" binaries downloaded earlier and renamed them respectively to match somehow the naming convention.
i.e. :  the `query-engine` file become `query-engine-rhel-openssl-1.0.x`

Then I've uploaded those binaries to my shared hosting where the binaries automatically downloaded by Prisma are.

```
// uploaded all four binaries here :
~/nodevenv/myapp/14/lib/node_modules/@prisma/engines/
```

Then I've changed the rights on those files to match the rights granted by default on the files automatically downloaded by Prisma.

Next I've edited the `schema.prisma` file to provide the correct binary targets: 

```
generator client {
  provider = "prisma-client-js"
  binaryTargets = ["rhel-openssl-1.0.x"]
}
//Models...
```
Next I've edited the `.env` file to provide the "correct" path to the new binaries :

```
PRISMA_QUERY_ENGINE_BINARY="/home/your_username_here/nodevenv/prismabin/14/lib/node_modules/@prisma/engines/query-engine-rhel-openssl-1.0.x"

PRISMA_MIGRATION_ENGINE_BINARY="/home/your_username_here/nodevenv/prismabin/14/lib/node_modules/@prisma/engines/migration-engine-rhel-openssl-1.0.x"

PRISMA_INTROSPECTION_ENGINE_BINARY="/home/your_username_here/nodevenv/prismabin/14/lib/node_modules/@prisma/engines/introspection-engine-rhel-openssl-1.0.x"

PRISMA_FMT_BINARY="/home/your_username_here/nodevenv/prismabin/14/lib/node_modules/@prisma/engines/prisma-fmt-rhel-openssl-1.0.x"
```
Then run `npx prima generate` 

Here I saw something that I was not expecting to see, my comand got executed ! 

Since my shared hosting do not allow to create a DB outside of the tools provided by Cpanel for creating a DB I could not use the `npx prisma migrate dev --name init` so the `npx prisma db push --preview-feature` made the trick and a table (as defined in the schema.prisma file) was created.

Here I still got an error even if the table is created: 
```
Error: ENOENT: no such file or directory, open '/home/ziken/nodevenv/prismabin/14/lib/node_modules/prisma/query-engine-debian-openssl-1.1.x'
```
So I've copied the file required file from `node_modules/@prisma/engines`
where Prisma want it to be, droped the table in the DB and tried `npx prisma db push --preview-feature` again, this time no error but a warning which I find to be pretty funny : 
```
Warning: Your current platform `debian-openssl-1.1.x` is not included in your generator's `binaryTargets` configuration ["rhel-openssl-1.0.x"].
```
**Now I can start playing with Prisma ! I guess x")**

Thanks for reading this and I'll let you know if I face some troubles again ! 

Thank you all and keep up the good work <3

**Reply by janpio** [reply]
Wow, amazing detective work in getting the the paths to download these binaries!

I think you can simplify your process a bit:

1. Develop your app locally as you are doing, no problem at all
2. Add this `binaryTarget` to your schema file: `binaryTargets = ["native", "rhel-openssl-1.0.x"]`
3. Local development should still work fine after running `prisma generate`, but it will download both binaries: You should now see _both_ binaries (the wrong `query-engine-debian-openssl-1.1.x` and the working `query-engine-rhel-openssl-1.0.x` in multiple location in `node_modules`
4. Now upload your project to your hosting environment, install dependencies and run `prisma generate` there: the same binaries will also be downloaded here. Now you can make the CLI and Prisma Client make use these.
5. Create a `.env` file in your hosting environment that sets the 4 binary env vars you mentioned above to the correct paths with `rhel-openssl-1.0.x`.
6. If the file is not enough, you can set the appropriate env var before calling any `prisma` command: `PRISMA_MIGRATION_ENGINE_BINARY=... npx prisma migrate dev` - but in _theory_ this should not be necessary. Let me know if it is, that might be a bug that want to fix.
6. Profit (and use Prisma)!

Let me know if this works or if/where it goes wrong.

The key thing you did here was finding out which binaries actually work on your environment. The rest is a bit of knowledge about the innards of Prisma and how it does binary stuff - we should document that better (if it actually works as described by me).

**Reply by kenzeasy** [reply]
Make Prisma downloading those binaries locally for a non existent Cent OS 7 by editing the schema file `binaryTargets = ["rhel-openssl-1.0.x"]` and running `prisma generate` was one my first approach.

It download only the `engine-query` one plus the `node_modules/@prisma/engines` folder is not containing any of the "new downloaded" binaires. I can find the `rhel-query-binairy` under  `node_module/prisma` and `node_modules/.prisma/client/` but nothing new is showing up under the `/engines`.

But I've tried to follow what you suggest, I had to upload the `node_module` folder because without it, prisma refuse to execute anything except `init` so it would be impossible for me to download the `rhel-binary` from the shared hosting:
```
// .env
// The path is /prisma/ and not /@prisma/engines since there is no "rhel-binary" under /@prisma/engines/
PRISMA_QUERY_ENGINE_BINARY="/home/your_username/nodevenv/prismatest/14/lib/node_modules/prisma/query-engine-rhel-openssl-1.0.x"
```
```
// schema.prisma
generator client {
  provider = "prisma-client-js"
  binaryTargets = ["rhel-openssl-1.0.x"]
}
```
If I let the "native" statement on my `binaryTargets`, i.e. : `binaryTargets = ["native","rhel-openssl-1.0.x"]` the terminal throw me :
```
Error: Schema parsing
thread 'main' panicked at 'cannot spawn executor threads: Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/async-global-executor-2.0.2/src/init.rs:39:18
stack backtrace:
   0: rust_begin_unwind
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/panicking.rs:92:14
   2: core::option::expect_none_failed
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/option.rs:1300:5
   3: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   4: async_io::driver::block_on
   5: async_global_executor::reactor::block_on
   6: async_global_executor::init::init_with_config
   7: core::ops::function::FnOnce::call_once
   8: once_cell::imp::OnceCell<T>::initialize::{{closure}}
   9: once_cell::imp::initialize_inner
  10: once_cell::imp::OnceCell<T>::initialize
  11: async_std::task::builder::Builder::blocking
  12: query_engine::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
It remain quite by putting only `binaryTargets = ["rhel-openssl-1.0.x"]`

From there `prisma generate` get executed flawlessly, I've tried to run `prisma db push --feature-preview` but it ask for the `migration engine` which was not downloaded in my local setup. So back to getting them from `binaries.prisma.sh`

Am I missing something out of your instructions ?

**Reply by janpio** [reply]
I might have just gotten it wrong. Sorry I wasted your time. I will experiment with my (untested) instructions a bit later, and report back. Maybe we can also improve some things at Prisma to make this easier.

**Reply by kenzeasy** [reply]
Hello again guys, some updates : I was developing locally (everything is fine locally) and decided to try what I got so far on the shared hosting.

By changing the binaries as mentioned above I managed to make `prisma generate` or `prisma db push --preview-feature` working, prisma client get downloaded and my tables are written on the DB. So I guess _something_ is working.

But it is pretty much it. Every use of the prisma client inside a js file never succeed.
 
I followed again the [start from scratch tuto](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch-node-postgres) and tried to execute the  `node index.js` `prisma.findMany()` part but nothing is showing up. 
The script execute and return as if I pressed Enter on the CLI no empty array, no error, nothing.

So I've changed the tuto a bit to see wtf is going on : 
```
// index.js
const { PrismaClient } = require('@prisma/client')

const prisma = new PrismaClient()

async function main() {
 try{

    const allUsers = await prisma.user.findMany()

    console.log(allUsers)
  }
  catch(error){
    console.log(error)
  }
}

main()
  .catch((e) => {
    throw e
  })
  .finally(async () => {
    await prisma.$disconnect()
  })
```

From there the terminal give me 2 diffrent errors depending on if I'm inside a node virtual env or not :
```
//inside node virtual env this never resolve I had to ctrl+c

Invalid `prisma.user.findMany()` invocation:


  Query engine exited with code 255
PANIC in /root/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-timer-3.0.2/src/native/delay.rs:112:21
timer has gone away
    at cb (/home/ziken/nodevenv/fixPrisma/14/lib/node_modules/@prisma/client/runtime/index.js:34786:17)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async main (/home/ziken/fixPrisma/index.js:8:22) {
  clientVersion: '2.21.2',
  errorCode: undefined
}
```
```
// outside of node virtual env

PrismaClientInitializationError2 [PrismaClientInitializationError]:
Invalid `prisma.user.findMany()` invocation:


  Query engine exited with code 101

thread 'main' panicked at 'cannot spawn executor threads: Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/async-global-executor-2.0.2/src/init.rs:39:18
stack backtrace:
   0: rust_begin_unwind
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/panicking.rs:92:14
   2: core::option::expect_none_failed
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/option.rs:1300:5
   3: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   4: async_io::driver::block_on
   5: async_global_executor::reactor::block_on
   6: async_global_executor::init::init_with_config
   7: core::ops::function::FnOnce::call_once
   8: once_cell::imp::OnceCell<T>::initialize::{{closure}}
   9: once_cell::imp::initialize_inner
  10: once_cell::imp::OnceCell<T>::initialize
  11: async_std::task::builder::Builder::blocking
  12: query_engine::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
    at cb (/home/ziken/nodevenv/fixPrisma/14/lib/node_modules/@prisma/client/runtime/index.js:34786:17)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async main (/home/ziken/fixPrisma/index.js:8:22) {
  clientVersion: '2.21.2',
  errorCode: undefined
}
```

I've seen some stuff that might be related to my issue [here](https://github.com/prisma/prisma/discussions/6070) and [here](https://github.com/prisma/prisma/discussions/4592)

I've already tried :
```
//index.js
const prisma = new PrismaClient({
// tried to set this to true or false
  __internal:{
    useUds: false
    
  }
} as any)
```
It changes nothing :(

More infos : 

- I've put some console.log inside my index.js file to see what part gets executed or not everything after the `await` line is never reach

- I've tried to install and run prisma outside of the node virtual env provided by the shared hosting but it gets even strangier since Prisma download another set of binaries, `debian-1.0.x` if I remember correctly.
I did not tried to reproduce what I have already done (downloading binaries from `binaries.prisma.sh`) with those since I want a web server and my shared hosting won't let me have one outside of the virtual node env.

- I know that my shared hosting is using [Phusion Passenger - Reverse port binding](https://www.phusionpassenger.com/library/indepth/nodejs/reverse_port_binding.html) but I don't know if this is have anything to do with my troubles ... 

As I said : I do want to make Prisma work :D

But I don't know what I have to do to make it behave as smooth as locally ^^ I guess I'm discovering the joy of deploying an app.

Thank you for reading any help is welcome <3

**Reply by janpio** [reply]
> But I don't know what I have to do to make it behave as smooth as locally ^^

The easiest solution will be to use another, more "common" host for your Node app.

To make it work on this host though, where it in theory can also work and possibly should, we might need to do some more investigation.

Can you maybe open an issue where you present your last state with all the necessary information to take a look at this independently?

**Reply by kenzeasy** [reply]
Another update from my journey : 

**Prisma is running flawlessly** (so far) **on my shared hosting** 

I can query my DB retrieve and write data inside it from my front-end. I must admit I'm pretty excited to dig deeper!
All I did is stop treating my shared hosting as my localhost (trying to run `node index.js` or using the terminal like I would locally to check if Prisma is doing his magic) and drop my Prisma queries inside an Express server through Cpanel NodeJS App installer (Yup, it was a desperate move). 

I still face sometimes some issues when running `prisma generate` which throw me some errors about resources being not available and things like that but it gets executed at some point, I'm done thinking I want to code.

My guess is that all the layers of security, virtualisation, network/resources management provided by the shared hosting is blocking Prisma from running inside the shared hosting terminal  so it must be used with an already existing "virtual web server app". I don't know if this make sense but to me it worked.

Anyway thanks again for reading this and huge shout out to @janpio for answering to my posts and providing a very much needed support to me. Give that man a raise <3

To the Prisma Team and all the contributors, maintainers : Change nothing you're amazing people thank you all for providing such a great tool. Keep up the good work.

**Reply by rtkay123** [reply]
> Another update from my journey :
> 
> **Prisma is running flawlessly** (so far) **on my shared hosting**
> 
> I can query my DB retrieve and write data inside it from my front-end. I must admit I'm pretty excited to dig deeper!
> All I did is stop treating my shared hosting as my localhost (trying to run `node index.js` or using the terminal like I would locally to check if Prisma is doing his magic) and drop my Prisma queries inside an Express server through Cpanel NodeJS App installer (Yup, it was a desperate move).
> 
> I still face sometimes some issues when running `prisma generate` which throw me some errors about resources being not available and things like that but it gets executed at some point, I'm done thinking I want to code.
> 
> My guess is that all the layers of security, virtualisation, network/resources management provided by the shared hosting is blocking Prisma from running inside the shared hosting terminal so it must be used with an already existing "virtual web server app". I don't know if this make sense but to me it worked.
> 
> Anyway thanks again for reading this and huge shout out to @janpio for answering to my posts and providing a very much needed support to me. Give that man a raise <3
> 
> To the Prisma Team and all the contributors, maintainers : Change nothing you're amazing people thank you all for providing such a great tool. Keep up the good work.

I was following your steps earlier but I'm hitting a hurdle I can't seem to get past. Are you available to help me step through it?

**Reply by kenzeasy** [reply]
I don't know if I can do much but sure tell me if there is anything I can do to help.

**Reply by rtkay123** [reply]
> I don't know if I can do much but sure tell me if there is anything I can do to help.

So I think i'm using the same environment as you. CentOS with OpenSSL 1.0.x
So basically, I followed all your steps. I got the binary files with my commit hash, added them to the `node_modules/@prisma` folder. I then configured the `.env` file. However, when I run `npx prisma generate` it somehow looks for the `debian-openssl-1.1.x` binary files (the paths in the environment variables are ignored). Funny thing is, when I delete the `rhel-openssl-1.1.x` binaries, then I run `npx prisma generate` then the compiler complains saying it can't find the `rhel-openssl-1.0.x` files that I put in the `.env` file. But when I include them, it ignores them and uses the `debian-openssl-1.1.x` ones. 
Is there anywhere else I need to change? I've tried different paths for the .env file. In project root, in prisma folder, same thing.

**Reply by kenzeasy** [reply]
Disclamer : 
I had some issues with Prisma and my shared hosting provider when I sent query to my db the connections stayed open and lead to process / memory overload. 
Tried to set `$prisma.disconnect` at the end of each functions that used `prisma/client` it was kind of 'ok' but it was too much for me.
So for now I've switched to heroku where it works out of the box. 

You can run `npx prisma version` to check if prisma uses the correct provided path in your  ̀.env` file.
In my setup, the `.env` file is inside my project root folder where it lands automatically.

What I can think of right now are those steps: 

Have you edited the `schema.prisma` file and set the binary targets to match Cent OS 7 without providing "native" just the one for cent OS : 
```
// schema.prisma file

generator client {
  provider = "prisma-client-js"
  binaryTargets = ["rhel-openssl-1.0.x"]
}
//Models...
```

Don't delete automatically downloaded binaries by prisma, at first I was trying to 'force' prisma to use the provided binaries by editing the `.env` file and removing every single binaries that prisma downloaded on its own. Saying to prisma "Well there is no debian binaries now what you gonna do about it !?" It didn't work ^^ (to me at least) 

Have you edited the rights of your 'manually downloaded' binaries to match the rights of automatically downloaded ones ?

There is several places where prisma is looking for binaries : 
`node_modules/@prisma/engines/` : there you should have all four binaries
`node_modules/.prisma/client` 
`node_modules/prisma`  

Sometimes prisma asked me for debian binary but inside `.prisma/client` for example, just copy paste the one that comes automatically inside the folder that prisma complains about (idk if this make sense). 

If your database is created and the correct credentials are provided to `DATABASE_URL` inside  `.env` file try to run `npx prisma migrate` or `npx prisma db push` to see if even without `prisma client` installed you can still 'reach' your database.

Just saw that the [start from scratch tuto](https://www.prisma.io/docs/getting-started/setup-prisma/start-from-scratch-node-postgres) has been updated.

It states that `prisma client` is now downloaded manually as an exclusive step (which if I remember correctly was not my case when I played with it). 

It will run `prisma generate` but you'll get those new prisma folder (I guess) where you can add the 'correct binaries' too even if `prisma generate` fails.

I might start over with the changes that prisma made during last months to see how it behave with my shared hosting now...

**Reply by jackpeters667** [reply]
So now i'm getting 

```
Error: Command was killed with SIGABRT (Aborted): /home/spectorw/nodevenv/nodejsapps/api/10/lib/node_modules/@prisma/engines/query-engine-rhel-openssl-1.0.x --version
thread 'main' panicked at 'cannot spawn executor threads: Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/async-global-executor-2.0.2/src/init.rs:39:18
```

I think i'm just going to switch to Heroku as well. I've been stuck on this for like 4 days lol. Hopefully someone can come through with it. Or the folk at prisma upload a shared hosting deployment guide

**Reply by kenzeasy** [reply]
What was the command you executed to get this error ? 
I had similar error on my side too when running `npx prisma generate` sometimes it get executed sometimes it doesn't ...

**Reply by jackpeters667** [reply]
> What was the command you executed to get this error ?
> I had similar error on my side too when running `npx prisma generate` sometimes it get executed sometimes it doesn't ...

Yeah it was `npx prisma generate` I get it all the time though.

(Sorry, I was away.. just got back)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
